### PR TITLE
parameterized distance metrics; lint; format; tests

### LINKED ISSUE #4368: Add distance metric param to to redis vectorstore index
### Feature request

Redis vectorstore allows for three different distance metrics: `L2` (flat L2), `COSINE`, and `IP` (inner product). Currently, the  `Redis._create_index` method hard codes the distance metric to COSINE.

```py
def _create_index(self, dim: int = 1536) -> None:
        try:
            from redis.commands.search.field import TextField, VectorField
            from redis.commands.search.indexDefinition import IndexDefinition, IndexType
        except ImportError:
            raise ValueError(
                "Could not import redis python package. "
                "Please install it with `pip install redis`."
            )

        # Check if index exists
        if not _check_index_exists(self.client, self.index_name):
            # Constants
            distance_metric = (
                "COSINE"  # distance metric for the vectors (ex. COSINE, IP, L2)
            )
            schema = (
                TextField(name=self.content_key),
                TextField(name=self.metadata_key),
                VectorField(
                    self.vector_key,
                    "FLAT",
                    {
                        "TYPE": "FLOAT32",
                        "DIM": dim,
                        "DISTANCE_METRIC": distance_metric,
                    },
                ),
            )
            prefix = _redis_prefix(self.index_name)

            # Create Redis Index
            self.client.ft(self.index_name).create_index(
                fields=schema,
                definition=IndexDefinition(prefix=[prefix], index_type=IndexType.HASH),
            )
```

This should be parameterized.

### Motivation

I'd like to be able to use L2 distance metrics.

### Your contribution

I've already forked and made a branch that parameterizes the distance metric in `langchain.vectorstores.redis`:

```py
 def _create_index(self, dim: int = 1536, distance_metric: REDIS_DISTANCE_METRICS = "COSINE") -> None:
        try:
            from redis.commands.search.field import TextField, VectorField
            from redis.commands.search.indexDefinition import IndexDefinition, IndexType
        except ImportError:
            raise ValueError(
                "Could not import redis python package. "
                "Please install it with `pip install redis`."
            )

        # Check if index exists
        if not _check_index_exists(self.client, self.index_name):
            # Define schema
            schema = (
                TextField(name=self.content_key),
                TextField(name=self.metadata_key),
                VectorField(
                    self.vector_key,
                    "FLAT",
                    {
                        "TYPE": "FLOAT32",
                        "DIM": dim,
                        "DISTANCE_METRIC": distance_metric,
                    },
                ),
            )
            prefix = _redis_prefix(self.index_name)

            # Create Redis Index
            self.client.ft(self.index_name).create_index(
                fields=schema,
                definition=IndexDefinition(prefix=[prefix], index_type=IndexType.HASH),
            ) def _create_index(self, dim: int = 1536, distance_metric: REDIS_DISTANCE_METRICS = "COSINE") -> None:
        try:
            from redis.commands.search.field import TextField, VectorField
            from redis.commands.search.indexDefinition import IndexDefinition, IndexType
        except ImportError:
            raise ValueError(
                "Could not import redis python package. "
                "Please install it with `pip install redis`."
            )

        # Check if index exists
        if not _check_index_exists(self.client, self.index_name):
            # Define schema
            schema = (
                TextField(name=self.content_key),
                TextField(name=self.metadata_key),
                VectorField(
                    self.vector_key,
                    "FLAT",
                    {
                        "TYPE": "FLOAT32",
                        "DIM": dim,
                        "DISTANCE_METRIC": distance_metric,
                    },
                ),
            )
            prefix = _redis_prefix(self.index_name)

            # Create Redis Index
            self.client.ft(self.index_name).create_index(
                fields=schema,
                definition=IndexDefinition(prefix=[prefix], index_type=IndexType.HASH),
            )

...

  @classmethod
    def from_texts(
        cls: Type[Redis],
        texts: List[str],
        embedding: Embeddings,
        metadatas: Optional[List[dict]] = None,
        index_name: Optional[str] = None,
        content_key: str = "content",
        metadata_key: str = "metadata",
        vector_key: str = "content_vector",
        distance_metric: REDIS_DISTANCE_METRICS = "COSINE",
        **kwargs: Any,
    ) -> Redis:
        """Create a Redis vectorstore from raw documents.
        This is a user-friendly interface that:
            1. Embeds documents.
            2. Creates a new index for the embeddings in Redis.
            3. Adds the documents to the newly created Redis index.
        This is intended to be a quick way to get started.
        Example:
            .. code-block:: python
                from langchain.vectorstores import Redis
                from langchain.embeddings import OpenAIEmbeddings
                embeddings = OpenAIEmbeddings()
                redisearch = RediSearch.from_texts(
                    texts,
                    embeddings,
                    redis_url="redis://username:password@localhost:6379"
                )
        """
        redis_url = get_from_dict_or_env(kwargs, "redis_url", "REDIS_URL")

        if "redis_url" in kwargs:
            kwargs.pop("redis_url")

        # Name of the search index if not given
        if not index_name:
            index_name = uuid.uuid4().hex

        # Create instance
        instance = cls(
            redis_url=redis_url,
            index_name=index_name,
            embedding_function=embedding.embed_query,
            content_key=content_key,
            metadata_key=metadata_key,
            vector_key=vector_key,
            **kwargs,
        )

        # Create embeddings over documents
        embeddings = embedding.embed_documents(texts)

        # Create the search index
        instance._create_index(dim=len(embeddings[0]), distance_metric=distance_metric)

        # Add data to Redis
        instance.add_texts(texts, metadatas, embeddings)
        return instance
```

I'll make the PR and link this issue

### PR DESCRIPTION
**Pull request by EandrewJones**
# Parameterize Redis vectorstore index

Redis vectorstore allows for three different distance metrics: `L2` (flat L2), `COSINE`, and `IP` (inner product). Currently, the  `Redis._create_index` method hard codes the distance metric to COSINE.

I've parameterized this as an argument in the `Redis.from_texts` method -- pretty simple.

Fixes #4368 

## Before submitting

I've added an integration test showing indexes can be instantiated with all three values in the `REDIS_DISTANCE_METRICS` literal. An example notebook seemed overkill here. Normal API documentation would be more appropriate, but no standards are in place for that yet.

## Who can review?

Not sure who's responsible for the vectorstore module... Maybe @eyurtsev / @hwchase17 / @agola11 ?

### CODE REVIEW

File: tests/integration_tests/vectorstores/test_redis.py [hunk 1]
```diff
@@ -58,3 +58,37 @@ def test_redis_add_texts_to_existing() -> None:
     output = docsearch.similarity_search("foo", k=2)
     assert output == TEST_RESULT
     assert drop(TEST_INDEX_NAME)
+
+
+class TestRedisDistanceMetrics:
+    """Test using different distance metrics for new indices."""
```
Comment (eyurtsev) [inline tests/integration_tests/vectorstores/test_redis.py:64]: Hello, @EandrewJones! :wave:  This is awesome! 

How difficult would it be to expand the test to `k=2` and have a meaningful difference for the 2nd retrieved result between the distance metrics?

The main concern is that with `k=1`, the test doesn't verify that the distance metric works correctly, only that the vector store can be instantiated with the given distance metric value. So a future refactor, could accidentally break this feature.

**Comment by eyurtsev** [comment]
@hwchase17  this looks reasonable to me. I haven't worked as much with the vector stores, so maybe another set of :eye:  will be helpful. Also left a challenge for @EandrewJones  if he's up for it. :)

File: tests/integration_tests/vectorstores/test_redis.py (hunk 1 above)
Comment (EandrewJones) [reply tests/integration_tests/vectorstores/test_redis.py:64]: Sure, thing. I can put something together with `k=2` for each metric showing the outputs and scores are as expected.

Though if the distance metrics don't work correctly that would imply an issue with the RedisStack implementation. We cannot directly address that on the langchain side apart from maybe excluding metrics that don't work correctly.

Comment (EandrewJones) [reply tests/integration_tests/vectorstores/test_redis.py:64]: Got it--By ensuring the expected results for `k=2` are unique to each metric, then we can guard against someone accidentally re-hardcoding the distance metric somewhere.

Will implement.

Comment (EandrewJones) [reply tests/integration_tests/vectorstores/test_redis.py:64]: @eyurtsev Any chance you can explain to me why  if I add an explicit `pytest.failure()` into an integration test, it still passes?

Comment (EandrewJones) [reply tests/integration_tests/vectorstores/test_redis.py]: @eyurtsev I've updated the test to include assertions unique to each distance metric by leveraging the scores produced for the second document when `k=2`.

Simple but it should cover the above concern you raise.

File: tests/integration_tests/vectorstores/test_redis.py [hunk 2]
```diff
@@ -58,3 +62,44 @@ def test_redis_add_texts_to_existing() -> None:
     output = docsearch.similarity_search("foo", k=2)
     assert output == TEST_RESULT
     assert drop(TEST_INDEX_NAME)
+
+
+class TestRedisDistanceMetrics:
```
Comment (dev2049) [inline tests/integration_tests/vectorstores/test_redis.py:67]: out of curiosity, what's value of having a class (instead of standalone test methods like above)

Comment (EandrewJones) [reply tests/integration_tests/vectorstores/test_redis.py:67]: Hey Davis.  Good question. Initially was reusing some logic across each test method in the class which justified it, but I ended up dropping it. I'm happy to go back to standalone functions if that's your preference.

In general, there are numerous benefits to using classes, however:

Organizing integration tests into a class instead of individual functions when using pytest can offer several advantages:

1. Logical grouping: Putting related integration tests into a class provides a clear logical grouping of test cases based on functionality or components under test. This makes it easier to understand the purpose of the tests, maintain them, and find specific tests when needed.

2. Code organization: Classes allow you to structure your test code in a more organized way, making it easier to read and maintain. You can define common utility methods or helper functions within the class that can be used by multiple test methods, reducing code duplication.

3. Test fixtures and setup/teardown methods: Pytest allows you to define setup and teardown methods for a class, which will be executed before and after each test method within the class. This can be useful for setting up and cleaning up any shared resources or state needed by multiple tests in the class, such as database connections, test data, or other dependencies.

4. Test parameterization: Pytest supports parameterized test methods within a class, which can reduce code duplication by allowing you to define a single test method and run it with different input parameters. This can be helpful for testing different scenarios or edge cases in your integration tests.

5. Test discovery and execution: By organizing tests into classes, you can more easily control test discovery and execution. You can run an entire class of tests or only specific test methods within a class, providing greater flexibility when running your tests.

Comment (EandrewJones) [reply tests/integration_tests/vectorstores/test_redis.py:68]: I'll go back to standalone. Convention is important until there's an overwhelming case for why we should change it. Eventually, I think the codebase's complexity will grow to a point where more within-file grouping is necessary to minimize repetition and improve organization.

Two other pieces of food for thought:

1. Test surface area grows faster than the feature space.
2. Targeted tests (re: 1+2+5 above) 

For example, all I do in this feature is add a single parameter to a class method but it results in _at least_ 3 test cases. So now there's a 1:3 ratio from feature to tests. With a class, I maintain a 1-1 mapping. Personally, I like 1-1 logical mappings. The src code files and the tests files also adhere to a 1-1 mapping right now. 

As far targeted runs, the one other nice thing I forgot to mention is I can then scope my tests to run just one class within a file.
